Add Follow-Focus plugin

l3nkz opened the pull request:
When this plugin is enabled, the input focus will follow the mouse
pointer. Hence, when moving the mouse and typing afterwards, the typed
keys will be send to the view that is currently under the cursor.
However, this will not change the stacking order of the views.

This behavior is similar to i3/sway or other X-based DEs.

Implements request from #16

soreau commented on metadata/follow-focus.xml:
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0"?>
+<wayfire>
+  <plugin name="follow-focus">
```
Plugin xml files use tabs, not spaces. See the other xml files for reference.

soreau commented on metadata/meson.build:
```diff
@@ -3,3 +3,4 @@ install_data('autorotate-iio.xml', install_dir: wayfire.get_variable(pkgconfig:
 install_data('background-view.xml', install_dir: wayfire.get_variable(pkgconfig: 'metadatadir'))
 install_data('bench.xml', install_dir: wayfire.get_variable(pkgconfig: 'metadatadir'))
 install_data('join-views.xml', install_dir: wayfire.get_variable(pkgconfig: 'metadatadir'))
+install_data('follow-focus.xml', install_dir: wayfire.get_variable(pkgconfig: 'metadatadir'))
```
I've been trying to keep the meson.build files containing the plugin lists in alphabetical order.

l3nkz commented:
> Perhaps we can add output-focus-follows-mouse to this plugin too. If we do, we'd probably want to make two bool options, one to enable view-focus and the other for output-focus. I can add it later if you don't feel like adding it here.
Yes, this is a great suggestion. I will add it to this plugin and update this pull request.

l3nkz commented:
I added the focus follow output feature, fixed your comments and also added the options to change the behavior of the plugin.

I am currently thinking about implementing a movement threshold. Especially with touchpads, it happens quite often that I accidentally "touch" the touchpad while typing and thereby change the window focus if the pointer is still on top of a different window.

Another problem, that I can't figure out yet is that if I for example close vim after changing the active view via the mouse pointer, the old window gets activated again (aka the one on top of the stack) maybe I need to restack the views as well when I set the view active.

ammen99 commented on src/follow-focus.cpp:
```diff
@@ -0,0 +1,84 @@
+    void change_view()
+    {
+        auto view = wf::get_core().get_cursor_focus_view();
+        wf::get_core().set_active_view(view);
```
You should use `wf::get_core().focus_view()` instead of `set_active_view`. The former will also raise the view to the front and focus the view's output.

l3nkz replied:
No, I just wanted to change the keyboard focus without restacking the views. This is what many old X-based DE used to do.

ammen99 replied:
Then focus won't be restored to the last view, that's how things work. You could try to keep a list of focused views in the plugin if you want to ...

ammen99 replied:
Btw, are you using simple-tile for all views? I don't think not-raising makes much sense with floating views (why would I type in a window which is 90% covered by another ??)

l3nkz replied:
Hm, usually, if I want to raise the window to the front I can just simply click on it. I use the plugin's functionality really for this use case where I want keyboard focus on another view without raising it. For example when I read papers and take notes, I usually have this work flow. In this case the pdf-viewer has ca. 80% of my screen and in addition to that I have a terminal with vim open (in the background) where I take notes …
But I don't have very hard feelings about the actual behavior. If you think that raising the window makes more sense, I can change this.

soreau replied:
It might make sense to have two options in this case, raise-on-click and focus-follows-mouse, since raise and focus are separate.

l3nkz replied:
Raise-on-click is afaik already provided by the core, isn't it?

soreau replied:
The point would be to not raise if the option is disabled. Maybe it's not the greatest choice of option name.

l3nkz replied:
Just to make sure, I understand you correctly: You suggest to always raise the views on top when the pointer hovers over the window and add an option to change this behavior so that the window only gets keyboard focus but isn't raised on top?

soreau replied:
>You should use wf::get_core().focus_view() instead of set_active_view. The former will also raise the view to the front and focus the view's output.

>I just wanted to change the keyboard focus without restacking the views. This is what many old X-based DE used to do.

I am saying to make these two behaviors to be selected with a toggle option.

l3nkz replied:
I think it is not possible to raise a view (using focus_view()) without giving it keyboard focus. Or did I miss something? Hence, I would think only one option makes sense, namely this one:
> always raise the views on top when the pointer hovers over the window and add an option to change this behavior so that the window only gets keyboard focus but isn't raised on top?

soreau replied:
Yes, I think you're right.

ammen99 replied:
> I think it is not possible to raise a view (using focus_view()) without giving it keyboard focus. 

This is perfectly possible, look at `workspace-manager.hpp`. However, when something triggers refocusing, that front view will be refocused.

l3nkz replied:
After having a closer look at `workspace-manager.hpp`, I don't think that its functionality can help me. At the moment `focus_view` does what is necessary when I want the window to be raised and if not `set_active_view` does the correct thing as well. How to fix the refocusing issue and that fast-switcher works again as expected, has to be solved differently. Both assume that stacking order in the workspace is equal to the activation order. However, this is an assumption that I break. Maybe adding a note to the option is enough for now?

ammen99 commented:
> Another problem, that I can't figure out yet is that if I for example close vim after changing the active view via the mouse pointer, the old window gets activated again (aka the one on top of the stack) maybe I need to restack the views as well when I set the view active.

See my comment, you are setting the keyboard focus only, but you should also raise the view on top, or at least that's what I assume you want to do?

Otherwise, if you don't want to raise the view below the mouse, you will run into this problem - because when the view is closed, wayfire will restore the focus to the topmost view.

l3nkz commented:
> Otherwise, if you don't want to raise the view below the mouse, you will run into this problem - because when the view is closed, wayfire will restore the focus to the topmost view.

Maybe my explanation of the situation wasn't the best. I have a terminal window open (the view that was activated by the plugin by hovering over it) and inside the terminal I have vim open. The mentioned refocus happens every time when I close vim inside the terminal without closing the view.

However, I think about changing the plugin a little bit again. Using the idle_call causes the refocus to happen after every little pointer movement. The result is, that if I move over multiple windows, the focus changes from one to another. However, my favorite behavior would be to actually wait until the pointer stops moving and only activate the view where the pointer finishes moving. Unfortunately,  I am not too familiar with the wayfire interface. Currently, I think about using the wl_timer feature and start a timer call upon pointer movement (with 2ms timeout or something like this) that gets rescheduled (disconnected and scheduled again) if there is more movement. And the timer callback than actually changes the focus. If I do it like this, implementing the threshold should be easy as well. I only have to remember the "start" position and check whether the pointer moved enough. What do you think?

soreau commented:
> Currently, I think about using the wl_timer feature and start a timer call upon pointer movement (with 2ms timeout or something like this) that gets rescheduled (disconnected and scheduled again) if there is more movement. And the timer callback than actually changes the focus. If I do it like this, implementing the threshold should be easy as well. I only have to remember the "start" position and check whether the pointer moved enough. What do you think?

You could add a `focus_delay` option which is the number of ms to wait before focusing. If the cursor position is updated before this times out, reset the timeout. Else, disconnect the timeout.

ammen99 commented:
What @soreau said, and yes, the timer interface looks like it can do the trick.

I will have to look at the code again to see what causes refocusing. As I said, core is pretty much hardcoded to focus the topmost view ...

l3nkz commented:
> What @soreau said, and yes, the timer interface looks like it can do the trick.

Ok, I will have a closer look and update the PR accordingly. Thank you guys for your feedback!

ammen99 commented:
I looked a bit through the code, if you run Wayfire with `-v` (or `-d` if you are running latest master), do you see some log messages when the view in front is again refocused?

I suspect that a layer-shell client triggers a refocus of some layer, in which case, you'd see `focusing layer ...`

l3nkz commented:
> I looked a bit through the code, if you run Wayfire with `-v` (or `-d` if you are running latest master), do you see some log messages when the view in front is again refocused?
> 
> I suspect that a layer-shell client triggers a refocus of some layer, in which case, you'd see `focusing layer ...`

I tried it and don't get such a message. I get the following messages when I open vim inside a terminal, and close it again:

```
DD 12-05-20 14:26:45.767 - [xwayland/xwm.c:830] XCB_CREATE_NOTIFY (4194305)
EE 12-05-20 14:26:45.767 - [wayfire-0.4.0/src/view/xwayland.cpp:231] new xwayland surface (null) class: (null) instance: (null)
DD 12-05-20 14:26:53.447 - [xwayland/xwm.c:844] XCB_DESTROY_NOTIFY (4194305)
```
I am a little bit confused, because I use termite as my terminal. Why do I get debug output from xwayland?

ammen99 commented:
@l3nkz The messages you posted are quite far apart (looking at the timestamps), so which one happens at the time the focus changes?

If you have a mind to further debug this and see what causes it: I believe the problem is that `output::refocus` gets called somehow. The code is https://github.com/WayfireWM/wayfire/blob/master/src/output/output.cpp#L51

It would be nice if you can add logging to that function, to see if it gets called as I suspect, and if yes, then add a call to `wf::print_trace(true)` to see what calls it.

l3nkz commented:
> The messages you posted are quite far apart (looking at the timestamps), so which one happens at the time the focus changes?

The last one happened when I closed vim. The other two happened when I opened vim.

> If you have a mind to further debug this and see what causes it: I believe the problem is that output::refocus gets called somehow. The code is https://github.com/WayfireWM/wayfire/blob/master/src/output/output.cpp#L51
It would be nice if you can add logging to that function, to see if it gets called as I suspect, and if yes, then add a call to wf::print_trace(true) to see what calls it.

I also suspected output::refocus to cause this behavior. However, I didn't have time yet to debug this. I will have a closer look tomorrow or on Friday.

wb9688 commented:
I noticed the following bugs:

- When I open a menu in Firefox, it closes as soon as I hover on it.
- Sometimes the keyboard focus will be on a window different from the window that will affect e.g. Win+Up (default key binding for `grid`).

I also noticed the following behavior that's different from what I would personally expect:

- When my mouse cursor enters an empty area, the window that was previously isn't focused anymore.
- When I e.g. use Alt+Tab (default key binding for `switcher`) and then move my mouse cursor, the focus is back on the original window were my mouse cursor is. The focus shouldn't be changed until my mouse cursor moves over the focused window and then on another window.

ammen99 commented:
> * When I e.g. use Alt+Tab (default key binding for `switcher`) and then move my mouse cursor, the focus is back on the original window were my mouse cursor is. The focus shouldn't be changed until my mouse cursor moves over the focused window and then on another window.

Sure this is expected? Sounds backward to me.

wb9688 commented:
@ammen99: That's what _I_ expect, yeah. Try in Sway ;)

xaiki commented:
FWIW I rebased this here: 
https://github.com/xaiki/wayfire-plugins-extra/tree/follow-focus

wb9688 commented:
@ammen99: Apparently Openbox has a different behavior from Sway: when opening a window or doing Alt+Tab, that window is focused until my mouse moves over a different window. On Sway, the focused window only changes once my mouse has first been on the focused window. On Wayfire, the focused window changes whenever the mouse is moved.
